crates/lib/src/http.rs

```diff
@@ -0,0 +1,545 @@
+//!
+//! The types here are all mirrors of various types within the [`http`] crate.
+//! That crate's types don't have stable representations or `pub`lic interiors,
+//! so we're forced to define our own representation for the SATS serialization.
+//! These types are that representation.
+//!
+//! To preserve extensibility and compatibility, all types defined here should be
+//! a `pub` wrapper struct with a single private `inner` field,
+//! which is an `enum` whose last variant holds either `Box<str>`, `String`, `Box<[T]>` or `Vec<T>` for any type `T`.
+//! Using an enum allows us to add additional variants while preserving the BSATN encoding passed across the WASM boundary,
+//! and including a variant with a variable-length type
+//! allows us to add other variants with variable-length types while preserving the BFLATN layout stored in table pages.
+//! (It's unlikely that any of these types will end up stored in a table, but better safe than sorry.)
+//!
+//! Users aren't intended to interact with these types, except [`Body`], [`Timeout`] and [`Error`].
+//! Our user-facing APIs should use the [`http`] crate's types directly, and convert to and from these types internally.
+
+use spacetimedb_sats::{time_duration::TimeDuration, SpacetimeType};
+
+/// Represents an HTTP request which can be made from a procedure running in a SpacetimeDB database.
+///
+/// Construct instances of this type by converting from [`http::Request`].
+#[derive(Clone, SpacetimeType)]
+#[sats(crate = crate)]
+pub struct Request {
+    inner: HttpRequest,
+}
+
+impl<T: Into<Body>> From<http::Request<T>> for Request {
+    fn from(req: http::Request<T>) -> Request {
+        let (
+            http::request::Parts {
+                method,
+                uri,
+                version,
+                headers,
+                mut extensions,
+                ..
+            },
+            body,
+        ) = req.into_parts();
+
+        let timeout = extensions.remove::<Timeout>();
+        if !extensions.is_empty() {
+            log::warn!("Converting HTTP `Request` with unrecognized extensions");
+        }
+        Request {
+            inner: HttpRequest::V0(HttpRequestV0 {
+                body: body.into(),
+                method: method.into(),
+                headers: headers.into(),
+                timeout,
+                uri: uri.to_string(),
+                version: version.into(),
+            }),
+        }
+    }
+}
+
+impl From<Request> for http::Request<Body> {
+    fn from(req: Request) -> http::Request<Body> {
+        let Request {
+            inner:
+                HttpRequest::V0(HttpRequestV0 {
+                    body,
+                    method,
+                    headers,
+                    timeout,
+                    uri,
+                    version,
+                }),
+        } = req
+        else {
+            unreachable!("`HttpRequest::NonExhausitve` pseudo-variant encountered");
```

Comment:
A user could theoretically pass a `NonExhaustive` to an abi call, causing a crash.